dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlHttpApiModule.java

```diff
@@ -40,6 +51,16 @@
 @ConfiguresHttpApi
 public class DimensionDataCloudControlHttpApiModule extends HttpApiModule<DimensionDataCloudControlApi> {
 
+   @Override
+   protected void installLocations() {
+      super.installLocations();
+      bind(RegionIdToURISupplier.class).to(RegionsToApiEndpoints.class).in(Scopes.SINGLETON);
+      bind(ZoneIdsSupplier.class).to(ZoneIdsFromRegionIdToZoneIdsValues.class).in(Scopes.SINGLETON);
+      bind(RegionIdsSupplier.class).to(RegionIdsFromRegionIdToURIKeySet.class).in(Scopes.SINGLETON);
```

Comment:
Remove this binding, as it is the one causing the circular dependency.

The `RegionsToApiEndpoints` supplier that provides the endpoint for each region uses the `@Region Supplier<Set<String>>` (the `RegionIdsSupplier`). but with this binding it was being configured to get the regions from the `Map<String, Supplier<URI>>` generated by the `RegionIdToURISupplier`, thus the circular dependency.

The default implementation of the `RegionIdsSupplier` just populates the list of regions based on the value of the `jclouds.regions` property, which is what we want here, so removing the binding should leave all good and the regions properly configured.

Reply:
With this bind removed the locations look properly populated. Tested with:
```java
public static void main(String[] args) {
      try (ApiContext<DimensionDataCloudControlApi> ctx =
              ContextBuilder.newBuilder("dimensiondata-cloudcontrol")
                 .credentials("foo", "bar") // Not needed. We won't call the API
                 .build()) {
         Supplier<Set<? extends Location>> locations = ctx.utils().injector()
               .getInstance(Key.get(new TypeLiteral<Supplier<Set<? extends Location>>>() {
               }, Memoized.class));
         
         for (Location l : locations.get()) {
            System.out.println(l);
         }
      }
   }
```
Prints:
```
{scope=REGION, id=na, description=na, parent=dimensiondata-cloudcontrol, iso3166Codes=[NA]}
{scope=REGION, id=eu, description=eu, parent=dimensiondata-cloudcontrol, iso3166Codes=[EU]}
{scope=REGION, id=au, description=au, parent=dimensiondata-cloudcontrol, iso3166Codes=[AU]}
{scope=REGION, id=mea, description=mea, parent=dimensiondata-cloudcontrol, iso3166Codes=[AF]}
{scope=REGION, id=ap, description=ap, parent=dimensiondata-cloudcontrol, iso3166Codes=[AP]}
{scope=REGION, id=canada, description=canada, parent=dimensiondata-cloudcontrol, iso3166Codes=[CA]}
{scope=ZONE, id=NA9, description=NA9, parent=na, iso3166Codes=[NA]}
{scope=ZONE, id=NA12, description=NA12, parent=na, iso3166Codes=[NA]}
{scope=ZONE, id=EU6, description=EU6, parent=eu, iso3166Codes=[EU]}
{scope=ZONE, id=EU7, description=EU7, parent=eu, iso3166Codes=[EU]}
{scope=ZONE, id=EU8, description=EU8, parent=eu, iso3166Codes=[EU]}
{scope=ZONE, id=AU9, description=AU9, parent=au, iso3166Codes=[AU]}
{scope=ZONE, id=AU10, description=AU10, parent=au, iso3166Codes=[AU]}
{scope=ZONE, id=AU11, description=AU11, parent=au, iso3166Codes=[AU]}
{scope=ZONE, id=AF3, description=AF3, parent=mea, iso3166Codes=[AF]}
{scope=ZONE, id=AP4, description=AP4, parent=ap, iso3166Codes=[AP]}
{scope=ZONE, id=AP5, description=AP5, parent=ap, iso3166Codes=[AP]}
{scope=ZONE, id=CA2, description=CA2, parent=canada, iso3166Codes=[CA]}
```

Reply:
Excellent, now the structure becomes very clear to me.